File: apps/web/src/lib/credentials.ts
// Web credential store. Defaults to in-memory (cleared on reload) so the bearer
// token and channel binding never touch disk; the password is never stored.
// A sessionStorage tier is offered for convenience within a tab session. The
// account Secret Key is intentionally NOT persisted here (ADR-012): the browser
// is a weaker keystore than an OS keychain, so the user re-enters it or pastes a
// Setup Code on a fresh tab.

import type { Session } from "./account";

export interface StoredCredentials {
  serverURL: string;
  email: string;
  session: Session;
}

const KEY = "tock.session.v1";

export interface CredentialStore {
  save(c: StoredCredentials): void;
  load(): StoredCredentials | null;
  clear(): void;
}

/** In-memory store (default). Cleared on page reload. */
export class MemoryStore implements CredentialStore {
  private value: StoredCredentials | null = null;
  save(c: StoredCredentials) {
    this.value = c;
  }
  load() {
    return this.value;
  }
  clear() {
    this.value = null;
  }
}

/** sessionStorage-backed store (per-tab, weaker; opt-in). */
export class SessionStorageStore implements CredentialStore {
  save(c: StoredCredentials) {
    sessionStorage.setItem(KEY, JSON.stringify(c));
  }
  load(): StoredCredentials | null {
    const raw = sessionStorage.getItem(KEY);
    return raw ? (JSON.parse(raw) as StoredCredentials) : null;
  }
  clear() {
    sessionStorage.removeItem(KEY);
  }
}

File: crates/tock-account/src/credentials.rs
//! Portable account credentials and a storage trait every client edge
//! implements (OS keyring on CLI, Keychain on Apple, `IndexedDB` on web).
//!
//! The password is intentionally absent — only the Secret Key, the SRP
//! session token, and its channel binding are retained, so a captured store
//! still cannot derive the URK without the user's password.

use serde::{Deserialize, Serialize};

/// Persisted account credentials. Sensitive fields (Secret Key, bearer token)
/// belong in a platform secret store; never serialise a password here.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct AccountCredentials {
    /// Sign-in server URL.
    pub server_url: String,
    /// Account email / login.
    pub username: String,
    /// Server-assigned account id.
    pub account_id: String,
    /// Emergency-Kit Secret Key string (`A4-…`).
    pub secret_key: String,
    /// Hex bearer token for `Authorization: Bearer`.
    pub bearer_token: String,
    /// Hex channel-binding tag for `X-Tock-Channel-Binding`.
    pub channel_binding: String,
    /// Absolute session expiry (Unix seconds).
    pub expires_at: i64,
}

impl AccountCredentials {
    /// Whether the session is expired relative to `now` (Unix seconds).
    #[must_use]
    pub const fn is_expired(&self, now: i64) -> bool {
        self.expires_at <= now
    }
}

/// A platform credential store. Implementations perform the actual I/O.
pub trait CredentialStore {
    /// Concrete store error.
    type Error;

    /// Persist credentials, replacing any existing entry.
    ///
    /// # Errors
    /// Propagates the platform store error.
    fn save(&self, creds: &AccountCredentials) -> Result<(), Self::Error>;

    /// Load credentials, if present.
    ///
    /// # Errors
    /// Propagates the platform store error.
    fn load(&self) -> Result<Option<AccountCredentials>, Self::Error>;

    /// Remove stored credentials (logout).
    ///
    /// # Errors
    /// Propagates the platform store error.
    fn clear(&self) -> Result<(), Self::Error>;
}
